R/utils-olson.R

```diff
@@ -0,0 +1,414 @@
+#' @keywords internal
+#' @seealso OlsonNames()
+#' @examples
+#' abbreviate_olson("America/Chicago")
+#' abbreviate_olson("America/Chicago", minwidth = 20)
+#' abbreviate_olson("America/Chicago", dictionary = c(America = "USA"))
+#' abbreviate_olson(OlsonNames())
+#' @export
+#'
+abbreviate_olson <- function(tz, minwidth = 14L, dictionary = NULL, ...) {
+
+  # ensure timezone is a character vector
+  if (!rlang::is_character(tz)) {
+    stop("tz must be a character vector", call. = FALSE)
+  }
+
+  # if we cannot return at least one character, return empty string
+  if (minwidth < 1L) {
+    return("")
+  }
+
+  # ensure non-null dictionary is a character vector
+  if (!is.null(dictionary) && !rlang::is_character(dictionary)) {
+    stop("dictionary must be a character vector", call. = FALSE)
+  }
+
+  # merge dictionaries, favouring the user-supplied set of dictionaries
+  dictionary_merged <- c(dictionary, .dict_default)
+
+  # create data frame for abbreviations
+  df <- .abbreviate_olson_df(tz, minwidth, dictionary = dictionary_merged, ...)
+
+  # recompose timezones
+  .recompose_tz <- function(x) {paste(x, collapse = "/")}
+  df_agg <-
+    stats::aggregate(df["abbv_final"], by = df["tz"], FUN = .recompose_tz)
+
+  # pull columns out of the data frame to form vector of abbreviations
+  tz_abbv <- df_agg[["abbv_final"]]
+  names(tz_abbv) <- df_agg[["tz"]]
+
+  # send back in the same order we received
+  tz_abbv[tz]
+}
+
+.abbreviate_olson_df <- function(tz, minwidth = 14L, dictionary = NULL, ...){
+
+  # This function may be useful to call for debugging
+  #
+  # tz           `character`, one-or-more Olson timezones
+  # minwidth     `integer`, width to which the timezones are to be abbreviated
+  # dictionary   `character`, named vector to insist on abbreviations
+  # ...           other args passed onto `abbreviate()`
+  #
+  # returns `data.frame` with variables:
+  #
+  # index
+  # index_max
+  # tz
+  # component
+  # budget_initial
+  # abbv_dict
+  # budget_remaining
+  # budget_provisional
+  # budget_final
+  # abbv_final
+
+  # create data frame with one row for each component of each timezone,
+  # with variables:
+  #
+  # index
+  # index_max
+  # tz
+  # component
+  # budget_initial
+  #
+  df <- .decompose_tz(tz, minwidth)
+
+  # apply dictionary (adds variable abbbv_dict)
+  df$abbv_dict <- .abbv_dict(df$component, df$budget_initial, dictionary)
+
+  # treat the one-, two-, and three-component timezones together
+  df_by_index_max <- split(df, df$index_max)
+
+  # the function .abbrevaite by index loops through index and returns the
+  # additional variables:
+  #
+  # budget_remaining
+  # budget_provisional
+  # budget_final
+  # abbv_final
+  #
+  list_df_new <-
+    map(df_by_index_max, .abbreviate_by_index, minwidth = minwidth, ...)
+
+  # bind the data frames together
+  df_new <- do.call(rbind, list_df_new)
+
+  df_new
+}
+
+.decompose_tz <- function(tz, minwidth = 14L) {
```

Comment:
I find the leading dot a bit distracting. "Private" functions are fine as long as they're not exported.

Reply:
done in  8097e5d